sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -410,13 +417,22 @@ public Read withKeyRange(ByteKeyRange keyRange) {
      *
      * <p>Does not modify this object.
      */
-    public Read withKeyRanges(List<ByteKeyRange> keyRanges) {
+    public Read withKeyRanges(ValueProvider<List<ByteKeyRange>> keyRanges) {
       checkArgument(keyRanges != null, "keyRanges can not be null");
-      checkArgument(!keyRanges.isEmpty(), "keyRanges can not be empty");
-      for (ByteKeyRange range : keyRanges) {
-        checkArgument(range != null, "keyRanges cannot hold null range");
-      }
-      return toBuilder().setKeyRanges(keyRanges).build();
+      BigtableReadOptions bigtableReadOptions = getBigtableReadOptions();
+      return toBuilder()
+          .setBigtableReadOptions(bigtableReadOptions.toBuilder().setKeyRanges(keyRanges).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Read} that will read only rows in the specified ranges.
+     * Ranges must not overlap.
```

Comment:
nit: currently it seems like overlapping ranges lead to duplicate reads for any keys in the overlapping area, it would be good if either:
1) overlapping ranges lead to an error
2) overlapping ranges are correctly fixed up to not produce duplicates

Reply:
Agreed - personally I would prefer option 2. I think it's out of scope for this PR, though, but I'll be happy to do it as a follow-up. Do you know if there is already a JIRA issue for this?

Reply:
I'm not aware of a JIRA, you'll need to look. If you don't find one, file one.